src/ecs/components/rendering/texture.rs

```diff
@@ -1,48 +1,53 @@
 //! Graphical texture resource.
 
-use gfx::Factory;
-use gfx::format::{Formatted, SurfaceTyped};
+pub use gfx::texture::Kind;
 
-use asset_manager::{AssetLoader, Assets};
+use gfx::Factory;
+use gfx::CombinedError as GfxError;
 use gfx_device::gfx_types;
+
+use asset_manager::Asset;
+use engine::Context;
 use renderer;
 use renderer::target::ColorFormat;
 
-pub use gfx::texture::Kind;
-
 /// Handle to a texture resource.
 pub type Texture = renderer::Texture<gfx_types::Resources>;
 
 /// Loads raw texture data.
-pub struct TextureLoadData<'a> {
+pub struct TextureLoadData {
     /// Type of storage data being used.
     pub kind: Kind,
     /// Slice of slices with each inner slice representing an image/texture's
     /// pixels laid out contiguously.
-    ///
-    /// FIXME: Ew, this is gross. Maybe we could use type aliases to make this
-    /// a little more readable?
-    pub raw: &'a [&'a [<<ColorFormat as Formatted>::Surface as SurfaceTyped>::DataType]],
+    pub raw: Box<[Box<[[u8; 4]]>]>,
 }
 
-impl<'a> AssetLoader<Texture> for TextureLoadData<'a> {
-    /// # Panics
-    ///
-    /// Panics if factory isn't registered as loader.
-    fn from_data(assets: &mut Assets, data: TextureLoadData) -> Option<Texture> {
-        let factory = assets.get_loader_mut::<gfx_types::Factory>()
-            .expect("Couldn't retrieve factory.");
-        let tex_res_view =
-            match factory.create_texture_immutable::<ColorFormat>(data.kind, data.raw) {
-                Ok((_, tex_res_view)) => tex_res_view,
-                Err(_) => return None,
-            };
-        Some(renderer::Texture::Texture(tex_res_view))
-    }
+/// Texture data, the data format for the
+/// texture asset.
+pub enum TextureData {
+    /// Raw texture data
+    Raw(TextureLoadData),
+    /// A one color texture
+    Constant([f32; 4]),
 }
 
-impl AssetLoader<Texture> for [f32; 4] {
-    fn from_data(_: &mut Assets, color: [f32; 4]) -> Option<Texture> {
-        Some(renderer::Texture::Constant(color))
+impl Asset for Texture {
+    type Data = TextureData;
+    type Error = GfxError;
+
+    fn from_data(data: TextureData, context: &mut Context) -> Result<Self, GfxError> {
```

Comment:
> But I thought glx wouldn't support Vulkan yet?

Not yet, but working on it!

Anyway, I assume in the future we could split the `Factory` into a single-threaded and a parallel part. The first one would be used in `from_data` and the parallel in `import`. This would be pretty neat imo.